Issue #2590: Add overwrite flag to CSV copy, overwrite only if query completes

JimStam opened the pull request:
Implementation for issue #2590.

Add an `OVERWRITE` flag to `COPY TO` queries. If set to `FALSE`, the file will **not** be overwritten if the query produces an error. The default behaviour remains to overwrite the file. 

If the flag is set to `FALSE` the CSV data is kept on the global state and is not flushed to the file until the query is complete.

JimStam committed:
Add unit test

JimStam committed:
Create test_nullable_boolean.py

JimStam committed:
Added additional testing, implemented overwrite flag and option in csv

JimStam committed:
Correctly process headers

JimStam committed:
Merge branch 'delete-overwrite'

JimStam committed:
Formatting fixes

JimStam committed:
Delete test_nullable_boolean.py

pdet commented on src/include/duckdb/execution/operator/persistent/buffered_csv_reader.hpp:
```diff
@@ -99,6 +99,8 @@ struct BufferedCSVReaderOptions {
 	std::map<LogicalTypeId, StrpTimeFormat> date_format = {{LogicalTypeId::DATE, {}}, {LogicalTypeId::TIMESTAMP, {}}};
 	//! Whether or not a type format is specified
 	std::map<LogicalTypeId, bool> has_format = {{LogicalTypeId::DATE, false}, {LogicalTypeId::TIMESTAMP, false}};
+	//! Whether the file should be overridden when query fails
```
should this be overwritten instead of overridden?
Also, I think the option should prolly be named more on the line of:
stream_write

pdet reviewed (commented):
Hey @JimStam, Thanks for your first PR, I think it looks great!

Could you also add some tests where the flag is set and the query succeeds?

Otherwise, I'm not sure if I like the name overwrite, because it gives the feeling that it will append if it's false. 

I think the name should be somewhere in the line of stream writing the file, or only writing after query completion (i.e., a short name that sends this idea)

JimStam committed:
Merge branch 'duckdb:master' into master

JimStam committed:
Renamed overwrite to stream_write and added unit tests for succesful query

JimStam commented:
Hey @pdet. I've changed to name to `stream_write` as per your suggestion. I think that covers the functionality well! I have also added a test for a succesful query with and without a header. It revealed that the stream_write flag needed to be configured in the  `ParseBaseOption` too, so I have added it there as well.

pdet commented:
@hannesmuehleisen could you set up CI approval for Jim's PRs?

Mytherin commented:
Hey Jim, thanks a lot for the PR!

The general strategy here seems to be to materialize the entire CSV file in memory, and to only write out the CSV file if the entire query has succeeded.

While that will work to fix certain types of problems (e.g. the query failing during the write due to an exception being thrown), it might still leave the user with a partially written CSV file. This can happen if, for example, the power shuts off unexpectedly while the system is in the process of writing out the final CSV file to disk.

In addition, this strategy also has the disadvantages that (1) writing is slower, since there is an additional materialization step in-memory, and (2) writing can only be done for CSV files that fit in memory.

Perhaps a better strategy is to first write the file to a temporary file (e.g. `out.csv.tmp`). When the query succeeds, we can use `FileSystem::MoveFile` to overwrite the original file (i.e. `mv out.csv.tmp out.csv`). This has the advantage that the file can be stream written as-is, just to a different location, which means there is no speed penalty or requirement for the file to fit in memory. If the query fails, the user is left just with an auxiliary file (`out.csv.tmp`) while the original file is left in-tact.

This strategy does have one disadvantage, which is that the user must have disk space for both files to fit on disk. Since disk space is generally bigger than memory I imagine that will be less of a problem than the memory constraint, however.

Perhaps it would even be possible to do this generically in the `PhysicalCopyToFile` operator, so this behavior works not just for the CSV writer but also for the Parquet writer.

JimStam committed:
Use .tmp file

JimStam committed:
Merge branch 'duckdb:master' into master

JimStam committed:
Remove unneeded white space

JimStam committed:
Merge branch 'master' of https://github.com/JimStam/duckdb

JimStam committed:
Move .tmp

JimStam committed:
Add a function for Moving tmp file

JimStam commented:
Hey @Mytherin, thanks for the feedback!

I've changed my pull request to add a `.tmp` while writing and upon finalizing using `FileSystem::MoveFile` to remove it again.

I have tried to add it generically in the `PhysicalCopyToFile` operator but I am having trouble accessing the file name from there and adding the `.tmp`. Do you have any suggestion on how to do this? Otherwise I can also do the same thing I have done in the functions for writing to CSV, but then for Parquet. Although I agree it would be better to make it more generic.

Mytherin commented:
> Hey @Mytherin, thanks for the feedback!
> 
> I've changed my pull request to add a `.tmp` while writing and upon finalizing using `FileSystem::MoveFile` to remove it again.
> 
> I have tried to add it generically in the `PhysicalCopyToFile` operator but I am having trouble accessing the file name from there and adding the `.tmp`. Do you have any suggestion on how to do this? Otherwise I can also do the same thing I have done in the functions for writing to CSV, but then for Parquet. Although I agree it would be better to make it more generic.

The file path is known in the `CopyStatement`, which gets turned into a `LogicalCopyToFIle` node in `bind_copy.cpp`. Afterwards, the `LogicalCopyToFile` node gets turned into a `PhysicalCopyToFile` node in `plan_copy_to_file.cpp`. You could add the file path to both the `LogicalCopyToFile` and the `PhysicalCopyToFile`, and pass it through the layers, so that you have access to the file path in the `PhysicalCopyToFile`.

JimStam committed:
Add .tmp generically from PhysicalCopyToFile

JimStam committed:
Formatting and remove unneeded comment

JimStam commented:
Hi @Mytherin and @pdet 

I've added the `.tmp` generically in the `PhysicalCopyToFile` operator. I've also added tests for parquet file formats.

Mytherin commented on src/execution/physical_plan/plan_copy_to_file.cpp:
```diff
@@ -6,9 +6,10 @@ namespace duckdb {
 
 unique_ptr<PhysicalOperator> PhysicalPlanGenerator::CreatePlan(LogicalCopyToFile &op) {
 	auto plan = CreatePlan(*op.children[0]);
+	op.file_path += ".tmp";
```
How about the scenario where we write to a pipe (e.g. `/dev/stdout`)?

SQL:

```sql
copy (select 42) to '/dev/stdout'
```

Mytherin commented on src/execution/physical_plan/plan_copy_to_file.cpp:
```diff
@@ -6,9 +6,10 @@ namespace duckdb {
 unique_ptr<PhysicalOperator> PhysicalPlanGenerator::CreatePlan(LogicalCopyToFile &op) {
 	auto plan = CreatePlan(*op.children[0]);
+	op.file_path += ".tmp";
 	// COPY from select statement to file
```
Can we add an option for this behavior (e.g. the original `OVERWRITE`)? There are some reasons why you might not want this behavior (e.g. potential to leave an extra file behind, have to fit both files on disk). Perhaps defaulting to false is safest, although I can see the argument for defaulting to true as long as we correctly detect e.g. the pipe scenario.

pdet replied:
I'm not so sure if this flag makes that much sense. 

I would say that cleaning up  .tmp files of failed queries should be enough since this will always be garbage from the query perspective anyway.

The two files on disk would only exist during query execution, which I think should be perfectly fine to all but 0.001% of scenarios, no? Otherwise, the user should do some disk cleanup.

Mytherin replied:
The file will be left behind when the system crashes and can't clean up. Perhaps not enough of a problem, but definitely a downside. Since we need to be able to turn this functionality off anyway (for when writing to a pipe fs or to a httpfs directly) I think having this flag there makes sense. But I agree that perhaps it should default to true.

pdet replied:
well, a system crash sounds more like a reason to have .tmp file than an overwrite.
I see the point for turning it off though, having the default to true seems the way go to.

One last thing, overwrite might be a bad name, if you want to expose that to a user.

Mytherin replied:
That depends if the user cares about preserving the original file, which won’t always be the case. Perhaps “use_temp_file” is a better name?

Mytherin reviewed (commented):
Thanks for the updates! Looking good. Some more comments:

JimStam committed:
Add check for pipe

pdet commented:
@JimStam you have to merge ur branch with master and solve the conflicts

JimStam committed:
Added stat and ISFIFO call on LocalFileSystem

pdet committed:
Only do .tmp if is an already existing file

pdet committed:
Use tmp option for csv

pdet committed:
Adding USE_TMP_FILE as a more generic option

pdet committed:
Merge remote-tracking branch 'upstream/master'

pdet committed:
Windows complaining about linux specific code and .sqltest complaining about missing space

pdet committed:
Missing include to make windows happy

pdet committed:
ups

pdet committed:
SQLTest complaining about missing space

pdet committed:
Breaking pipe tests to windows and linux

pdet committed:
Format fix checks

pdet committed:
Trying to make windows 32 happy

Mytherin commented:
Could you have a look the CI failures on MingW?

```
2022-02-10T15:46:08.5111710Z D:/a/duckdb/duckdb/src/common/file_system.cpp:235:2: error: invalid preprocessing directive #elifdef; did you mean #ifdef?
2022-02-10T15:46:08.6191551Z   235 | #elifdef _WIN32
2022-02-10T15:46:08.7019010Z       |  ^~~~~~~
2022-02-10T15:46:08.7468929Z       |  ifdef
2022-02-10T15:46:08.7476915Z D:/a/duckdb/duckdb/src/common/file_system.cpp: In static member function 'static bool duckdb::FileSystem::IsFile(const string&)':
2022-02-10T15:46:08.7586135Z D:/a/duckdb/duckdb/src/common/file_system.cpp:230:55: error: no matching function for call to 'stat::stat(const char*, _stat32*)'
2022-02-10T15:46:08.7592495Z   230 |                         stat(filename.c_str(), &status);
2022-02-10T15:46:08.7593284Z       |                                                       ^
2022-02-10T15:46:08.7616988Z In file included from D:/a/_temp/msys64/mingw64/x86_64-w64-mingw32/include/wchar.h:413,
2022-02-10T15:46:08.7633908Z                  from D:/a/_temp/msys64/mingw64/include/c++/11.2.0/cwchar:44,
2022-02-10T15:46:08.7769224Z                  from D:/a/_temp/msys64/mingw64/include/c++/11.2.0/bits/postypes.h:40,
2022-02-10T15:46:08.7876466Z                  from D:/a/_temp/msys64/mingw64/include/c++/11.2.0/iosfwd:40,
2022-02-10T15:46:08.7899067Z                  from D:/a/_temp/msys64/mingw64/include/c++/11.2.0/bits/shared_ptr.h:52,
2022-02-10T15:46:08.7910793Z                  from D:/a/_temp/msys64/mingw64/include/c++/11.2.0/memory:77,
2022-02-10T15:46:08.7930459Z                  from D:/a/duckdb/duckdb/src/include/duckdb/common/constants.hpp:11,
2022-02-10T15:46:08.7932630Z                  from D:/a/duckdb/duckdb/src/include/duckdb/common/common.hpp:11,
2022-02-10T15:46:08.7933326Z                  from D:/a/duckdb/duckdb/src/include/duckdb/common/allocator.hpp:11,
2022-02-10T15:46:08.7933934Z                  from D:/a/duckdb/duckdb/src/common/allocator.cpp:1,
2022-02-10T15:46:08.7934561Z                  from D:/a/duckdb/duckdb/src/common/ub_duckdb_common.cpp:2:
2022-02-10T15:46:08.7935279Z D:/a/_temp/msys64/mingw64/x86_64-w64-mingw32/include/_mingw_stat64.h:40:10: note: candidate: 'stat::stat()'
2022-02-10T15:46:08.7935858Z    40 |   struct stat {
2022-02-10T15:46:08.7936317Z       |          ^~~~
2022-02-10T15:46:08.7936915Z D:/a/_temp/msys64/mingw64/x86_64-w64-mingw32/include/_mingw_stat64.h:40:10: note:   candidate expects 0 arguments, 2 provided
2022-02-10T15:46:08.7937834Z D:/a/_temp/msys64/mingw64/x86_64-w64-mingw32/include/_mingw_stat64.h:40:10: note: candidate: 'constexpr stat::stat(const stat&)'
2022-02-10T15:46:08.7938576Z D:/a/_temp/msys64/mingw64/x86_64-w64-mingw32/include/_mingw_stat64.h:40:10: note:   candidate expects 1 argument, 2 provided
2022-02-10T15:46:08.7939278Z D:/a/_temp/msys64/mingw64/x86_64-w64-mingw32/include/_mingw_stat64.h:40:10: note: candidate: 'constexpr stat::stat(stat&&)'
2022-02-10T15:46:08.7939902Z D:/a/_temp/msys64/mingw64/x86_64-w64-mingw32/include/_mingw_stat64.h:40:10: note:   candidate expects 1 argument, 2 provided
2022-02-10T15:46:08.7940495Z In file included from D:/a/duckdb/duckdb/src/common/ub_duckdb_common.cpp:14:
2022-02-10T15:46:08.7941086Z D:/a/duckdb/duckdb/src/common/file_system.cpp:39:26: error: '_S_IFMT' was not declared in this scope
2022-02-10T15:46:08.7941609Z    39 | #define S_ISREG(x) (((x)&_S_IFMT) == _S_IFREG)
2022-02-10T15:46:08.7941996Z       |                          ^~~~~~~
2022-02-10T15:46:08.7942480Z D:/a/duckdb/duckdb/src/common/file_system.cpp:39:26: note: in definition of macro 'S_ISREG'
2022-02-10T15:46:08.7942991Z    39 | #define S_ISREG(x) (((x)&_S_IFMT) == _S_IFREG)
2022-02-10T15:46:08.7943391Z       |                          ^~~~~~~
2022-02-10T15:46:08.7943922Z D:/a/duckdb/duckdb/src/common/file_system.cpp:39:38: error: '_S_IFREG' was not declared in this scope; did you mean 'S_ISREG'?
2022-02-10T15:46:08.7944484Z    39 | #define S_ISREG(x) (((x)&_S_IFMT) == _S_IFREG)
2022-02-10T15:46:08.7944912Z       |                                      ^~~~~~~~
2022-02-10T15:46:08.7945413Z D:/a/duckdb/duckdb/src/common/file_system.cpp:39:38: note: in definition of macro 'S_ISREG'
2022-02-10T15:46:08.7946497Z    39 | #define S_ISREG(x) (((x)&_S_IFMT) == _S_IFREG)
2022-02-10T15:46:08.7946912Z       |                                      ^~~~~~~~
2022-02-10T15:46:08.7947574Z D:/a/duckdb/duckdb/src/common/file_system.cpp:238:55: error: no matching function for call to 'stat::stat(const char*, stat*)'
2022-02-10T15:46:08.7948193Z   238 |                         stat(filename.c_str(), &status);
2022-02-10T15:46:08.7948593Z       |                                                       ^
2022-02-10T15:46:08.7949085Z In file included from D:/a/_temp/msys64/mingw64/x86_64-w64-mingw32/include/wchar.h:413,
2022-02-10T15:46:08.7949622Z                  from D:/a/_temp/msys64/mingw64/include/c++/11.2.0/cwchar:44,
2022-02-10T15:46:08.7950139Z                  from D:/a/_temp/msys64/mingw64/include/c++/11.2.0/bits/postypes.h:40,
2022-02-10T15:46:08.7950627Z                  from D:/a/_temp/msys64/mingw64/include/c++/11.2.0/iosfwd:40,
2022-02-10T15:46:08.7951135Z                  from D:/a/_temp/msys64/mingw64/include/c++/11.2.0/bits/shared_ptr.h:52,
2022-02-10T15:46:08.7951642Z                  from D:/a/_temp/msys64/mingw64/include/c++/11.2.0/memory:77,
2022-02-10T15:46:08.7952172Z                  from D:/a/duckdb/duckdb/src/include/duckdb/common/constants.hpp:11,
2022-02-10T15:46:08.7952722Z                  from D:/a/duckdb/duckdb/src/include/duckdb/common/common.hpp:11,
2022-02-10T15:46:08.7953270Z                  from D:/a/duckdb/duckdb/src/include/duckdb/common/allocator.hpp:11,
2022-02-10T15:46:08.7953792Z                  from D:/a/duckdb/duckdb/src/common/allocator.cpp:1,
2022-02-10T15:46:08.7954304Z                  from D:/a/duckdb/duckdb/src/common/ub_duckdb_common.cpp:2:
2022-02-10T15:46:08.7954879Z D:/a/_temp/msys64/mingw64/x86_64-w64-mingw32/include/_mingw_stat64.h:40:10: note: candidate: 'stat::stat()'
2022-02-10T15:46:08.7955331Z    40 |   struct stat {
2022-02-10T15:46:08.7955684Z       |          ^~~~
2022-02-10T15:46:08.7956195Z D:/a/_temp/msys64/mingw64/x86_64-w64-mingw32/include/_mingw_stat64.h:40:10: note:   candidate expects 0 arguments, 2 provided
2022-02-10T15:46:08.7956866Z D:/a/_temp/msys64/mingw64/x86_64-w64-mingw32/include/_mingw_stat64.h:40:10: note: candidate: 'constexpr stat::stat(const stat&)'
2022-02-10T15:46:08.7957563Z D:/a/_temp/msys64/mingw64/x86_64-w64-mingw32/include/_mingw_stat64.h:40:10: note:   candidate expects 1 argument, 2 provided
2022-02-10T15:46:08.7958257Z D:/a/_temp/msys64/mingw64/x86_64-w64-mingw32/include/_mingw_stat64.h:40:10: note: candidate: 'constexpr stat::stat(stat&&)'
2022-02-10T15:46:08.7958895Z D:/a/_temp/msys64/mingw64/x86_64-w64-mingw32/include/_mingw_stat64.h:40:10: note:   candidate expects 1 argument, 2 provided
2022-02-10T15:46:08.7959485Z In file included from D:/a/duckdb/duckdb/src/common/ub_duckdb_common.cpp:14:
2022-02-10T15:46:08.7960072Z D:/a/duckdb/duckdb/src/common/file_system.cpp:39:26: error: '_S_IFMT' was not declared in this scope
2022-02-10T15:46:08.7960569Z    39 | #define S_ISREG(x) (((x)&_S_IFMT) == _S_IFREG)
2022-02-10T15:46:08.7960977Z       |                          ^~~~~~~
2022-02-10T15:46:08.7961496Z D:/a/duckdb/duckdb/src/common/file_system.cpp:39:26: note: in definition of macro 'S_ISREG'
2022-02-10T15:46:08.7962026Z    39 | #define S_ISREG(x) (((x)&_S_IFMT) == _S_IFREG)
2022-02-10T15:46:08.7962408Z       |                          ^~~~~~~
2022-02-10T15:46:08.7962943Z D:/a/duckdb/duckdb/src/common/file_system.cpp:39:38: error: '_S_IFREG' was not declared in this scope; did you mean 'S_ISREG'?
2022-02-10T15:46:08.7963495Z    39 | #define S_ISREG(x) (((x)&_S_IFMT) == _S_IFREG)
2022-02-10T15:46:08.7963905Z       |                                      ^~~~~~~~
2022-02-10T15:46:08.7964374Z D:/a/duckdb/duckdb/src/common/file_system.cpp:39:38: note: in definition of macro 'S_ISREG'
2022-02-10T15:46:08.7964884Z    39 | #define S_ISREG(x) (((x)&_S_IFMT) == _S_IFREG)
2022-02-10T15:46:08.7965293Z       |                                      ^~~~~~~~
2022-02-10T15:46:08.7965859Z [28/218] Building CXX object src/common/serializer/CMakeFiles/duckdb_common_serializer.dir/ub_duckdb_common_serializer.cpp.obj
```

pdet committed:
mingw not happy with fs

pdet committed:
Wait apparently there were already FS type functions

pdet committed:
Revert "Wait apparently there were already FS type functions"

This reverts commit 2d9d6e75e0a99a342002a186e1ed8b48bdcb3cc7.

pdet committed:
I think i got this windows pipe detection right this time by actually checking the status size, but I've been wrong about this before

pdet committed:
Merge both file exists methods to one

Mytherin commented on src/common/local_file_system.cpp:
```diff
@@ -284,17 +285,26 @@ bool LocalFileSystem::DirectoryExists(const string &directory) {
-bool LocalFileSystem::FileExists(const string &filename) {
+bool LocalFileSystem::FileExists(const string &filename, bool empty_is_valid) {
 	if (!filename.empty()) {
+#if defined(_WIN32) && defined(__MINGW32__)
```
Windows (and MingW) use a different code path - this function is never called on Windows.

Mytherin commented on src/common/local_file_system.cpp:
```diff
@@ -284,17 +285,26 @@ bool LocalFileSystem::DirectoryExists(const string &directory) {
 			struct stat status;
 			stat(filename.c_str(), &status);
-			if (!(status.st_mode & S_IFDIR)) {
+			if ((status.st_size > 0 || empty_is_valid) && !(status.st_mode & S_IFDIR)) {
```
Is it not sufficient to check the `st_mode` and verify that it is not a pipe?

Mytherin commented on test/sql/copy/csv/test_copy.test:
```diff
@@ -4,175 +4,175 @@
-# test invalid UTF-8
-statement error
-COPY test FROM 'test/sql/copy/csv/data/test/invalid_utf.csv' DELIMITER '|';
+#statement ok
```
Commented out by accident?

pdet committed:
Only way I got Mark to agree with this pipe check on mingw

pdet committed:
Glob now breaks on named pipes

pdet committed:
I guess I also have to do the widestring shenanigans for mingw

Mytherin commented on src/common/local_file_system.cpp:
```diff
@@ -52,26 +52,32 @@ static void AssertValidFileFlags(uint8_t flags) {
-		stat(filename.c_str(), &status);
+	int wchars_num = MultiByteToWideChar(CP_UTF8, 0, filename.c_str(), -1, NULL, 0);
+	wchar_t *wpath = new wchar_t[wchars_num];
+	MultiByteToWideChar(CP_UTF8, 0, filename.c_str(), -1, wpath, wchars_num);
```
We have some helper functions for this (e.g. `WindowsUtil::UTF8ToUnicode`)

pdet committed:
use WindowsUtil::UTF8ToUnicode
